File: typescript/packages/var-runner/src/steps.ts
import { pathToFileURL } from 'node:url'
import { _resetBuilder, buildRegistry, contextFactory } from '@oselvar/var/registry'
import { findFiles } from '@oselvar/var-config'
import type { Registry } from '@oselvar/var-core'

export type LoadedSteps = {
  readonly registry: Registry
  readonly createContext: (stepFile: string) => unknown | Promise<unknown>
}

export async function loadSteps(
  stepGlobs: ReadonlyArray<string>,
  cwd: string,
): Promise<LoadedSteps> {
  _resetBuilder()
  for (const path of findFiles(cwd, stepGlobs)) {
    await import(pathToFileURL(path).href)
  }
  return { registry: buildRegistry(), createContext: contextFactory() }
}

File: ruby/packages/var-runner/lib/oselvar/var/runner/steps.rb
# frozen_string_literal: true

require 'oselvar/var'
require 'oselvar/var/registry'

module Oselvar
  module Var
    module Runner
      # The registry + per-file context factory built from loaded step files.
      LoadedSteps = Data.define(:registry, :create_context)

      module_function

      # Reset the accumulator, load (execute) every step file matching
      # +step_globs+ under +root+, and build the registry + context factory.
      def load_steps(step_globs, root)
        RegistryGlue.reset_builder
        files = []
        step_globs.each do |g|
          files.concat(Dir.glob(g, base: root).map { |rel| File.join(root, rel) })
        end
        files.select { |p| File.file?(p) }.uniq.sort.each { |path| load path }
        LoadedSteps.new(registry: RegistryGlue.build_registry, create_context: RegistryGlue.context_factory)
      end
    end
  end
end
